lib/handlebars/compiler/javascript-compiler.js

```diff
@@ -500,16 +499,53 @@ JavaScriptCompiler.prototype = {
     this.pushStackLiteral(omitEmpty ? 'undefined' : '{}');
   },
   pushHash: function() {
-    if (this.hash) {
-      this.hashes.push(this.hash);
+    if (this.hashPieces) {
+      this.hashes.push(this.hashPieces);
     }
-    this.hash = {values: [], types: [], contexts: [], ids: []};
+    this.hashPieces = [];
   },
   popHash: function() {
-    let hash = this.hash;
-    this.hash = this.hashes.pop();
+    let hashPieces = this.hashPieces;
+    this.hashPieces = this.hashes.pop();
+    this.hashPiece = this.hashPieces && this.hashPieces[this.hashPieces.length - 1];
```

Comment:
What use to just be stack of hash objects is now a stack of hash piece arrays, and `this.hashPiece` refers to the last (current) hashPiece, so we restore it to that last element unless we've popped off all the hash piece arrays.